File: typescript/src/hub.ts
import {
  DeliveryLog,
  type DeliveryCursor,
  type DeliveryEntry,
  type DeliveryRecovery,
} from './recovery.js';
import { SubscriptionRegistry, type SubscriptionStats } from './subscriptions.js';

export interface SubscriptionHubOptions {
  /** Application-owned identity for this delivery-log lifetime. */
  readonly epoch: string;
  readonly maxEntries?: number;
  readonly initialSequence?: number;
}

/** Pure delivery work for a transport adapter to execute. */
export interface DeliveryPlan<Stream, Connection, Payload> {
  readonly entry: DeliveryEntry<Stream, Payload>;
  readonly connections: readonly Connection[];
}

/**
 * Composes subscription indexes and ordered recovery without performing I/O.
 *
 * A transport adapter remains responsible for authorization, serialization, sending, retries,
 * and pruning connections that fail during an actual send.
 */
export class SubscriptionHub<Stream, Connection extends object, Payload> {
  readonly #subscriptions = new SubscriptionRegistry<Stream, Connection>();
  readonly #deliveries: DeliveryLog<Stream, Payload>;

  constructor(options: SubscriptionHubOptions) {
    this.#deliveries = new DeliveryLog(options);
  }

  subscribe(stream: Stream, connection: Connection): boolean {
    return this.#subscriptions.subscribe(stream, connection);
  }

  unsubscribe(stream: Stream, connection: Connection): boolean {
    return this.#subscriptions.unsubscribe(stream, connection);
  }

  removeConnection(connection: Connection): number {
    return this.#subscriptions.removeConnection(connection);
  }

  planDelivery(
    stream: Stream,
    streamVersion: number,
    payload: Payload,
  ): DeliveryPlan<Stream, Connection, Payload> {
    return {
      entry: this.#deliveries.append(stream, streamVersion, payload),
      connections: this.#subscriptions.subscribers(stream),
    };
  }

  recoverAfter(
    cursor: DeliveryCursor,
    authorizedStreams: ReadonlySet<Stream>,
  ): DeliveryRecovery<Stream, Payload> {
    return this.#deliveries.recoverAfter(cursor, authorizedStreams);
  }

  entries(): readonly DeliveryEntry<Stream, Payload>[] {
    return this.#deliveries.entries();
  }

  retainedStreams(): ReadonlySet<Stream> {
    return new Set(this.#deliveries.entries().map((entry) => entry.stream));
  }

  historySize(): number {
    return this.#deliveries.size();
  }

  latestCursor(): DeliveryCursor {
    return this.#deliveries.latestCursor();
  }

  subscriberCount(stream: Stream): number {
    return this.#subscriptions.subscriberCount(stream);
  }

  streamCount(): number {
    return this.#subscriptions.streamCount();
  }

  stats(): SubscriptionStats {
    return this.#subscriptions.stats();
  }
}

/**
 * Collapses an invalidation-style replay to the newest retained delivery for each stream.
 *
 * This helper is deliberately opt-in: applications whose payloads are non-idempotent deltas must
 * replay every entry instead.
 */
export function latestDeliveryPerStream<Stream, Payload>(
  entries: readonly DeliveryEntry<Stream, Payload>[],
): readonly DeliveryEntry<Stream, Payload>[] {
  const latest = new Map<Stream, DeliveryEntry<Stream, Payload>>();
  let epoch: string | undefined;
  for (const entry of entries) {
    epoch ??= entry.cursor.epoch;
    if (entry.cursor.epoch !== epoch) {
      throw new RangeError('deliveries from different epochs cannot be collapsed together');
    }
    const existing = latest.get(entry.stream);
    if (existing === undefined || entry.cursor.sequence > existing.cursor.sequence) {
      latest.set(entry.stream, entry);
    }
  }
  return [...latest.values()].sort((left, right) => left.cursor.sequence - right.cursor.sequence);
}

File: rust/src/hub.rs
use std::collections::{BTreeMap, HashSet};
use std::hash::Hash;

use crate::{
    DeliveryCursor, DeliveryEntry, DeliveryLog, DeliveryRecovery, RealtimeError,
    SubscriptionRegistry, SubscriptionStats,
};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct DeliveryPlan<Stream, Connection, Payload> {
    pub entry: DeliveryEntry<Stream, Payload>,
    pub connections: Vec<Connection>,
}

/// Composes subscription indexes and ordered recovery without performing I/O.
#[derive(Clone, Debug)]
pub struct SubscriptionHub<Stream, Connection, Payload> {
    subscriptions: SubscriptionRegistry<Stream, Connection>,
    deliveries: DeliveryLog<Stream, Payload>,
}

impl<Stream, Connection, Payload> SubscriptionHub<Stream, Connection, Payload>
where
    Stream: Clone + Eq + Hash + Ord,
    Connection: Clone + Ord,
    Payload: Clone,
{
    /// Creates an empty hub.
    ///
    /// # Errors
    ///
    /// Returns the same epoch and capacity errors as [`DeliveryLog::new`].
    pub fn new(epoch: impl Into<String>, max_entries: usize) -> Result<Self, RealtimeError> {
        Self::with_initial_sequence(epoch, max_entries, 0)
    }

    /// Creates an empty hub whose next delivery follows `initial_sequence`.
    ///
    /// # Errors
    ///
    /// Returns the same epoch and capacity errors as [`DeliveryLog::with_initial_sequence`].
    pub fn with_initial_sequence(
        epoch: impl Into<String>,
        max_entries: usize,
        initial_sequence: u64,
    ) -> Result<Self, RealtimeError> {
        Ok(Self {
            subscriptions: SubscriptionRegistry::new(),
            deliveries: DeliveryLog::with_initial_sequence(epoch, max_entries, initial_sequence)?,
        })
    }

    pub fn subscribe(&mut self, stream: Stream, connection: Connection) -> bool {
        self.subscriptions.subscribe(stream, connection)
    }

    pub fn unsubscribe(&mut self, stream: &Stream, connection: &Connection) -> bool {
        self.subscriptions.unsubscribe(stream, connection)
    }

    pub fn remove_connection(&mut self, connection: &Connection) -> usize {
        self.subscriptions.remove_connection(connection)
    }

    /// Appends a delivery and returns the connections a transport should notify.
    ///
    /// # Errors
    ///
    /// Returns [`RealtimeError::SequenceExhausted`] before mutation when exhausted.
    pub fn plan_delivery(
        &mut self,
        stream: Stream,
        stream_version: u64,
        payload: Payload,
    ) -> Result<DeliveryPlan<Stream, Connection, Payload>, RealtimeError> {
        let connections = self
            .subscriptions
            .subscribers(&stream)
            .into_iter()
            .cloned()
            .collect();
        let entry = self.deliveries.append(stream, stream_version, payload)?;
        Ok(DeliveryPlan { entry, connections })
    }

    #[must_use]
    pub fn recover_after(
        &self,
        cursor: &DeliveryCursor,
        authorized_streams: &HashSet<Stream>,
    ) -> DeliveryRecovery<Stream, Payload> {
        self.deliveries
            .recover_after(cursor, Some(authorized_streams))
    }

    #[must_use]
    pub fn entries(&self) -> impl ExactSizeIterator<Item = &DeliveryEntry<Stream, Payload>> {
        self.deliveries.entries()
    }

    #[must_use]
    pub fn retained_streams(&self) -> HashSet<Stream> {
        self.deliveries
            .entries()
            .map(|entry| entry.stream.clone())
            .collect()
    }

    #[must_use]
    pub fn history_len(&self) -> usize {
        self.deliveries.len()
    }

    #[must_use]
    pub fn latest_cursor(&self) -> DeliveryCursor {
        self.deliveries.latest_cursor()
    }

    #[must_use]
    pub fn subscriber_count(&self, stream: &Stream) -> usize {
        self.subscriptions.subscriber_count(stream)
    }

    #[must_use]
    pub fn stream_count(&self) -> usize {
        self.subscriptions.stream_count()
    }

    #[must_use]
    pub fn stats(&self) -> SubscriptionStats {
        self.subscriptions.stats()
    }
}

/// Keeps only the newest delivery for each stream, ordered by final delivery sequence.
///
/// This is opt-in because non-idempotent delta payloads must replay every entry.
/// # Errors
///
/// Returns [`RealtimeError::MixedDeliveryEpoch`] rather than comparing sequence numbers from
/// unrelated delivery-log lifetimes.
pub fn latest_delivery_per_stream<Stream, Payload>(
    entries: impl IntoIterator<Item = DeliveryEntry<Stream, Payload>>,
) -> Result<Vec<DeliveryEntry<Stream, Payload>>, RealtimeError>
where
    Stream: Clone + Ord,
{
    let mut latest = BTreeMap::<Stream, DeliveryEntry<Stream, Payload>>::new();
    let mut epoch = None::<String>;
    for entry in entries {
        match &epoch {
            None => epoch = Some(entry.cursor.epoch.clone()),
            Some(epoch) if epoch != &entry.cursor.epoch => {
                return Err(RealtimeError::MixedDeliveryEpoch);
            }
            Some(_) => {}
        }
        let should_replace = latest
            .get(&entry.stream)
            .is_none_or(|existing| entry.cursor.sequence > existing.cursor.sequence);
        if should_replace {
            latest.insert(entry.stream.clone(), entry);
        }
    }
    let mut entries = latest.into_values().collect::<Vec<_>>();
    entries.sort_by_key(|entry| entry.cursor.sequence);
    Ok(entries)
}
